torch/distributed/pipelining/_PipelineStage.py

```diff
@@ -819,3 +820,399 @@ def __init__(
         # Get my pipe info
         pipe_info = pipe.info()
         super().__init__(stage_module, stage_index, pipe_info, device, group)
+
+
+# Manual PipelineStage functions and definition
+
+METADATA_TENSOR_LEN = 100
+PLACEHOLDER_VAL = -1
+
+
+def create_empty_tensors(
```

Comment:
as a lower priority thing than the shape inference, I'd like to replace this with use of pytree if that makes sense. The upside would be not having to write a helper function at all, and at the same time getting other data structures (like a dict of tensors) for free.

The  downsides could be- perf isn't great for pytree, but it wont matter if we only use it during initialization; we would be supporting more types of inputs, but then we'd also need to define what types of inputs are _not_ supported and guard against those.